temporalio/testing/workflow.py

```diff
@@ -0,0 +1,503 @@
+"""Workflow test environment."""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+import os
+import platform
+import shutil
+import socket
+import stat
+import subprocess
+import tarfile
+import tempfile
+import urllib.request
+import zipfile
+from contextlib import asynccontextmanager, contextmanager
+from datetime import datetime, timedelta, timezone
+from io import BytesIO
+from pathlib import Path
+from typing import (
+    Any,
+    AsyncIterator,
+    Callable,
+    Iterable,
+    Iterator,
+    Mapping,
+    Optional,
+    Type,
+    Union,
+    cast,
+)
+
+import google.protobuf.empty_pb2
+
+import temporalio.api.testservice.v1
+import temporalio.client
+import temporalio.common
+import temporalio.converter
+import temporalio.exceptions
+import temporalio.types
+import temporalio.worker
+
+logger = logging.getLogger(__name__)
+
+
+class WorkflowEnvironment:
+    """Workflow environment for testing workflows.
+
+    Most developers will want to use the static :py:meth:`start_time_skipping`
+    to start a test server process that automatically skips time as needed.
+
+    This environment is an async context manager, so it can be used with
+    ``async with`` to make sure it shuts down properly. Otherwise,
+    :py:meth:`shutdown` can be manually called.
+
+    To use the environment, simply use the :py:attr:`client` on it.
+
+    Workflows invoked on the workflow environment are automatically configured
+    to have ``assert`` failures fail the workflow with the assertion error.
+    """
+
+    @staticmethod
+    def from_client(client: temporalio.client.Client) -> WorkflowEnvironment:
```

Comment:
Nice!

Reply:
Wait until you see a future PR where I will have `start_embedded` static function that's gonna download/run Temporalite like it does the test server. I wrestled with myself on whether that's valuable, but considering that I'm gonna do that for my own Python tests, I think it is.

Reply:
Yes please, but please put this in Core so we can have it for TS and newer SDKs.